terraform/deployments/govuk-publishing-platform/signon_bearer_tokens.tf

```diff
@@ -15,6 +15,179 @@ resource "random_password" "signon_admin_password" {
   special = true
 }
 
+#
+# SecretsManager Rotation Lambda
+#
+
+locals {
+  bearer_token_lambda_name = "bearer_token_rotater-${local.workspace}"
+}
+
+data "archive_file" "bearer_token_rotater" {
+  type        = "zip"
+  source_file = "${path.module}/../../../lambdas/signon_bearer_token_rotater.rb"
+  output_path = "${path.module}/../../../lambdas/signon_bearer_token_rotater.zip"
+}
+
+resource "aws_lambda_function" "bearer_token" {
+  function_name = local.bearer_token_lambda_name
+  role          = aws_iam_role.lambda_execution_role.arn
+  handler       = "${local.bearer_token_lambda_name}.handler"
+
+  filename         = data.archive_file.bearer_token_rotater.output_path
+  source_code_hash = data.archive_file.bearer_token_rotater.output_base64sha256
+
+  runtime = "ruby2.7"
+
+  vpc_config {
+    subnet_ids         = local.private_subnets
+    security_group_ids = [aws_security_group.signon_lambda.id]
+  }
+
+  environment {
+    variables = {
+      ADMIN_PASSWORD_KEY = aws_secretsmanager_secret.signon_admin_password.arn
+      # TODO: Should be HTTPS
+      SIGNON_API_URL = "http://${module.signon.virtual_service_name}/api/v1"
+    }
+  }
+
+  # As recommended in Terraform docs
+  depends_on = [
+    aws_iam_role_policy_attachment.lambda_logs,
+    aws_iam_role_policy_attachment.vpc,
+    aws_cloudwatch_log_group.bearer_token,
+  ]
+
+  tags = merge(
+    local.additional_tags,
+    {
+      Name = "${local.bearer_token_lambda_name}-${var.govuk_environment}-${local.workspace}"
+    },
+  )
+}
+
+resource "aws_cloudwatch_log_group" "bearer_token" {
+  name              = "/aws/lambda/${local.bearer_token_lambda_name}"
+  retention_in_days = 90
+}
+
+# Adds a trust policy to the lambda function.
+resource "aws_lambda_permission" "allow_secretsmanager" {
```

Comment:
If I understand correctly: the execution role determines what the Lambda can do, whereas these `aws_lambda_permission`s determine what other services can do to the Lambda.

Reply:
That's a good question. I was a little confused by the permissions.

There are two sets of permissions since they do different things:

- the lambda execution role. This enables the lambda to take actions, such as log to CloudWatch, rotate secrets, and interact with the VPC.
- the lambda [resource-based policy](https://docs.aws.amazon.com/lambda/latest/dg/access-control-resource-based.html). The [lambda_permission](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/lambda_permission) allows other AWS services (Secrets Manager in this case) to invoke the function